registration/include/pcl/registration/impl/correspondence_estimation.hpp

```diff
@@ -126,45 +127,86 @@ CorrespondenceEstimation<PointSource, PointTarget, Scalar>::determineCorresponde
-  pcl::Correspondence corr;
-  unsigned int nr_valid_correspondences = 0;
+  std::vector<pcl::Correspondences> per_thread_correspondences(num_threads_);
+  for(auto& corrs: per_thread_correspondences) {
+    corrs.reserve(2 * indices_->size() / num_threads_);
+  }
 
   // Check if the template types are the same. If true, avoid a copy.
   // Both point types MUST be registered using the POINT_CLOUD_REGISTER_POINT_STRUCT
   // macro!
   if (isSamePointType<PointSource, PointTarget>()) {
     // Iterate over the input set of source indices
-    for (const auto& idx : (*indices_)) {
+    #pragma omp parallel for \
+      default(none) \
+      shared(tree_, indices_, max_dist_sqr, per_thread_correspondences) \
+      firstprivate(index, distance) \
+      num_threads(num_threads_)
+    for (int i = 0; i < static_cast<int>(indices_->size()); i++) {
+      const auto& idx = (*indices_)[i];
       tree_->nearestKSearch((*input_)[idx], 1, index, distance);
       if (distance[0] > max_dist_sqr)
         continue;
 
+      pcl::Correspondence corr;
       corr.index_query = idx;
       corr.index_match = index[0];
       corr.distance = distance[0];
-      correspondences[nr_valid_correspondences++] = corr;
+      per_thread_correspondences[omp_get_thread_num()].push_back(corr);
     }
   }
   else {
-    PointTarget pt;
-
     // Iterate over the input set of source indices
-    for (const auto& idx : (*indices_)) {
+    #pragma omp parallel for \
+      default(none) \
+      shared(tree_, indices_, max_dist_sqr, per_thread_correspondences) \
+      firstprivate(index, distance) \
+      num_threads(num_threads_)
+    for (int i = 0; i < static_cast<int>(indices_->size()); i++ ) {
+      const auto& idx = (*indices_)[i];
+
       // Copy the source data to a target PointTarget format so we can search in the
       // tree
+      PointTarget pt;
       copyPoint((*input_)[idx], pt);
 
       tree_->nearestKSearch(pt, 1, index, distance);
       if (distance[0] > max_dist_sqr)
         continue;
 
+      pcl::Correspondence corr;
       corr.index_query = idx;
       corr.index_match = index[0];
       corr.distance = distance[0];
-      correspondences[nr_valid_correspondences++] = corr;
+      per_thread_correspondences[omp_get_thread_num()].push_back(corr);
     }
   }
-  correspondences.resize(nr_valid_correspondences);
+
+  if(num_threads_ == 1) {
+    correspondences.swap(per_thread_correspondences.front());
+  } else {
+    int nr_correspondences = std::accumulate(
+      per_thread_correspondences.begin(),
+      per_thread_correspondences.end(),
+      0,
+      [](const auto sum, const auto& corr) { return sum + corr.size(); }
+    );
+    correspondences.resize(nr_correspondences);
+
+    // Merge per-thread correspondences while keeping them ordered
+    auto insert_loc = correspondences.begin();
+    for(const auto& corrs : per_thread_correspondences) {
+      std::copy(corrs.begin(), corrs.end(), insert_loc);
```

Comment:
```suggestion
      const auto new_insert_loc = std::move(corrs.begin(), corrs.end(), insert_loc);
```

We can then replace the L203 and L206 with this variable (bigger win is the use of `move` over `copy`

Reply:
Done